apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex

```diff
@@ -0,0 +1,274 @@
+
+      c = ^end_char, acc ->
+        acc = %{acc | paren_count: acc.paren_count - 1, text: [acc.text | [c]]}
+
+        if acc.paren_count <= 0 do
+          {:halt, acc}
+        else
+          {:cont, acc}
+        end
+
+      c, acc ->
+        {:cont, %{acc | text: [acc.text | [c]]}}
+    end)
+    |> Map.get(:text)
+    |> IO.iodata_to_binary()
+  end
+
+  defp get_pipe_call(line, col, head, current, tail) do
+    pipe_right = do_get_function_call(tail, "(", ")")
+
+    reversed_head = String.reverse(head)
+
+    {pipe_left, _, _} =
+      for <<x::binary-size(1) <- reversed_head>>, reduce: {"", false, false} do
+        {acc, _, true} ->
+          {acc, true, true}
+
+        {acc, has_passed_through_non_whitespace, _should_halt} ->
+          is_whitespace = String.match?(x, ~r/\s/)
+
+          cond do
+            is_whitespace and has_passed_through_non_whitespace ->
+              {x <> acc, true, true}
+
+            is_whitespace and not has_passed_through_non_whitespace ->
+              {x <> acc, false, false}
+
+            not is_whitespace and (x == "|" or x == ">") ->
+              {x <> acc, false, false}
+
+            true ->
+              {x <> acc, true, false}
+          end
+      end
+
+    pipe_left =
+      if String.contains?(pipe_left, ")") do
+        get_function_call_before(head)
+      else
+        pipe_left
+      end
+
+    pipe_call = pipe_left <> current <> pipe_right
+
+    {line_offset, tail_length} = pipe_left |> String.reverse() |> count_newlines_and_get_tail()
+
+    start_line = line - line_offset
+
+    start_col =
+      if line_offset != 0 do
+        head
+        |> String.trim_trailing(pipe_left)
+        |> String.split("\n")
+        |> Enum.at(-1, "")
+        |> String.length()
+      else
+        col - tail_length
+      end
+
+    {line_offset, tail_length} = (current <> pipe_right) |> count_newlines_and_get_tail()
+
+    end_line = line + line_offset
+
+    end_col =
+      if line_offset != 0 do
+        tail_length
+      else
+        col + tail_length - 1
+      end
+
+    {:ok, pipe_call, range(start_line, start_col, end_line, end_col)}
+  end
+
+  defp get_function_call_before(head) do
+    call_without_function_name =
+      head
+      |> String.reverse()
+      |> do_get_function_call(")", "(")
+      |> String.reverse()
+
+    function_name =
+      head
+      |> String.trim_trailing(call_without_function_name)
+      |> get_function_name_from_tail()
+
+    function_name <> call_without_function_name
+  end
+
+  defp get_function_name_from_tail(s) do
+    s
+    |> String.reverse()
+    |> String.graphemes()
+    |> Enum.reduce_while([], fn c, acc ->
+      if String.match?(c, ~r/\s/) do
+        {:halt, acc}
+      else
+        {:cont, [c | acc]}
+      end
+    end)
+    |> IO.iodata_to_binary()
+  end
+
+  defp count_newlines_and_get_tail(s) do
+    for <<c::binary-size(1) <- s>>, reduce: {0, 0} do
+      {count, tail} ->
+        if c == "\n" do
```

Comment:
Add /r/n handling

Reply:
@lukaszsamson doesn't this already cover that case? The main thing here is to get the number of lines and any remaining characters after "\n"

Reply:
Also disregard this. For covering "\r" some refactoring here is needed

Reply:
\r\n is a bit tricky to do right. LSP treats \n, \r and \n\r as newlines but elixir parser only supports \n and \r\n. On the other hand elixir allows \r inside strings, charlists and heredocs. I once needed something similar in one of the tests https://github.com/elixir-lsp/elixir-ls/blob/794ed49a571f7943d478ef0a87ed94ab07b30c67/apps/language_server/test/source_file_test.exs#L67

Reply:
Yeah, these are showing to be the hardest part of the code haha